File: 操作系统/Windows API/Windows 程序设计/第6版/CSharp/Chapter11/Petzold.ProgrammingWindows6.Chapter11/Petzold.ProgrammingWindows6.Chapter11/NamedColor.cs
using System.Collections.Generic;
using System.Reflection;
using Windows.UI;

namespace Petzold.ProgrammingWindows6.Chapter11
{
    public class NamedColor
    {
        static NamedColor()
        {
            List<NamedColor> colorList = new List<NamedColor>();
            IEnumerable<PropertyInfo> properties = typeof(Colors).GetTypeInfo().DeclaredProperties;

            foreach (PropertyInfo property in properties)
            {
                NamedColor namedColor = new NamedColor
                {
                    Name = property.Name,
                    Color = (Color)property.GetValue(null)
                };

                colorList.Add(namedColor);
            }

            All = colorList;
        }

        public static IEnumerable<NamedColor> All { private set; get; }

        public string Name { private set; get; }

        public Color Color { private set; get; }
    }
}

File: 操作系统/Windows API/Windows 程序设计/第6版/CPlusPlus/Chapter11/PetzoldProgrammingWindows6Chapter11/PetzoldProgrammingWindows6Chapter11/NamedColor.cpp
#include "pch.h"
#include "NamedColor.h"

using namespace PetzoldProgrammingWindows6Chapter11;

using namespace Platform;
using namespace Platform::Collections;
using namespace Windows::Foundation::Collections;
using namespace Windows::UI;

// C# assembly
using namespace ReflectionHelper;

IIterable<NamedColor^>^ NamedColor::all = nullptr;

IIterable<NamedColor^>^ NamedColor::All::get()
{
    if (all == nullptr)
    {
       Vector<NamedColor^>^ colorVector = ref new Vector<NamedColor^>();

        // Implemented in C# assembly ReflectionHelper
        TypeInformation^ typeInfo = ref new TypeInformation(Colors::typeid);
        IIterable<PropertyInformation^>^ properties = typeInfo->DeclaredProperties;

        for each (PropertyInformation^ property in properties)
        {
            NamedColor^ namedColor = ref new NamedColor();
            namedColor->name = property->Name;
            namedColor->color = (Windows::UI::Color)property->GetValue(nullptr);
            colorVector->Append(namedColor);
        }

        all = colorVector;
    }
    return all;
}
